### PR TITLE
Support multi-threading

### LINKED ISSUE #36: support multithread debugging

### PR DESCRIPTION
**Pull request by LeszekSwirski**
Fixes #36. Not everything works perfectly, in particular non-varobj local don't have a well-defined thread, but it's a good base to build on. I'm not 100% sure about the logic behind VAR_HANDLES_START, so the current array size increase may be overkills

Also includes some odds-and-ends fixing parsing and reference printing.

### CODE REVIEW

File: src/backend/mi_parse.ts [hunk 1]
```diff
@@ -205,8 +205,10 @@ export function parseMI(output: string): MINode {
 		let oldContent = output;
 		let canBeValueList = output[0] == '[';
 		output = output.substr(1);
-		if (output[0] == '}' || output[0] == ']')
+		if (output[0] == '}' || output[0] == ']') {
+			output = output.substr(1); // ] or }
 			return [];
+		}
```
Comment (WebFreak001) [inline src/backend/mi_parse.ts]: uh any reason for this change? Could you include some test what this is supposed to fix or what didn't work in the tests file please

Comment (LeszekSwirski) [reply src/backend/mi_parse.ts]: I don't remember the exact command that caused it, but it was to do with an mi command returning {} followed by other data, where the other data was getting lost

Comment (LeszekSwirski) [reply src/backend/mi_parse.ts]: I've added an (artificial) test, which I can confirm did fail before this patch.

**Comment by WebFreak001** [comment]
ok while quickly going over the changes I didn't notice any major issues so I will do more extensive testing later with all supported debuggers (not sure if mago supports multithreading, but it's just a side product anyway to allow D people on windows to debug if they really want to).

I don't really know anything about async in js so I need to lookup if any thread safety issues can occur there or if it's just fancy promises.

Otherwise I don't know how I feel about all the missing semicolons and using `'` for strings longer than 1 character yet lol

But this seems like a good and useful PR, thanks

**Comment by LeszekSwirski** [comment]
async in JS is effectively equivalent to promise `then` chains, there's still no actual parallelism or threading. I can fix the semicolons and `'`s.

I've only tested with gdb so YMMV, single-threaded functionality shouldn't change with other compilers though I think.

**Comment by LeszekSwirski** [comment]
Gentle reminder.

**Comment by WebFreak001** [comment]
sorry was kind of busy with other stuff and didn't really feel like reviewing something large, I can do it now

**Comment by WebFreak001** [comment]
**GDB**
- [x] single threaded still works locally
- [x] multi threaded now works locally
- [x] single threaded still works over SSH
- [x] multi threaded now works over SSH
- [ ] single threaded still works over gdbserver
- [ ] multi threaded now works over gdbserver
- [ ] single threaded still works when attaching
- [ ] multi threaded now works when attaching
Exception:
```
TypeError: Cannot read property 'getThreads' of undefined
    at GDBDebugSession.threadsRequest (/tmp/code-debug/out/src/mibase.js:239:24)
    at GDBDebugSession.dispatchRequest (/tmp/code-debug/node_modules/vscode-debugadapter/lib/debugSession.js:398:22)
    at GDBDebugSession._handleData (/tmp/code-debug/node_modules/vscode-debugadapter/lib/protocol.js:97:38)
    at Socket.inStream.on (/tmp/code-debug/node_modules/vscode-debugadapter/lib/protocol.js:18:44)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at readableAddChunk (_stream_readable.js:178:18)
    at Socket.Readable.push (_stream_readable.js:136:10)
    at Pipe.onread (net.js:560:20): Error: TypeError: Cannot read property 'getThreads' of undefined
    at GDBDebugSession.threadsRequest (/tmp/code-debug/out/src/mibase.js:239:24)
    at GDBDebugSession.dispatchRequest (/tmp/code-debug/node_modules/vscode-debugadapter/lib/debugSession.js:398:22)
    at GDBDebugSession._handleData (/tmp/code-debug/node_modules/vscode-debugadapter/lib/protocol.js:97:38)
    at Socket.inStream.on (/tmp/code-debug/node_modules/vscode-debugadapter/lib/protocol.js:18:44)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at readableAddChunk (_stream_readable.js:178:18)
    at Socket.Readable.push (_stream_readable.js:136:10)
    at Pipe.onread (net.js:560:20)
    at file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:2644:530
    at Object.v [as _notify] (file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:140:302)
    at Object.enter (file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:143:568)
    at n.Class.derive._oncancel._run (file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:144:866)
    at n.Class.derive._oncancel._error (file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:144:383)
    at file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:2223:975
    at _.e.dispatch (file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:2225:291)
    at _.e.handleData (file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:2224:815)
    at Socket.<anonymous> (file:///opt/visual-studio-code/resources/app/out/vs/workbench/workbench.main.js:2223:835)
    at emitOne (events.js:96:13)
```

**LLDB**
- [x] single threaded still works locally
- [x] multi threaded now works locally.
minor issue: thread names say undefined 
- [ ] single threaded still works over SSH (breakpoints don't work)
- [ ] multi threaded now works over SSH (breakpoints don't work)
- [ ] single threaded still works when attaching
- [ ] multi threaded now works when attaching


**Mago-MI**
~- [ ] single threaded still works locally~ (can't test because I can't run a VM because I disabled AMD-V in my bios because it locks my CPU after some random amount of time)
I don't think testing anything else here makes sense, because it is very limited

But I supposed most of these issues were present already, I am going to check for completeness though. The only thing I see when testing here right now is therefore that undefined in the thread name of lldb.

**Comment by WebFreak001** [comment]
ok so the Thread exception when it fails to connect is new, but gdbserver still doesn't work.

I did find one issue though: breakpoints show as verified in the broken lldb ssh connection now, they were unverified like this  before

**Comment by WebFreak001** [comment]
I can merge this as it is now because it is a good start, but I would prefer it if there were some `// TODO` comments where I just commented what behaviour changed

**Comment by LeszekSwirski** [comment]
Thanks for the review, I've added some defensive ifs and TODOs. Regarding the breakpoints being verified, I think there's some semantic difference between breakpoints being verified, and being enabled. Since I get the "done" response back from lldb, I assume they're verified as far as lldb is concerned (modulo lldb mi bugs), but they might not actually be enabled because of various reasons. Regardless, there's a new callback-based breakpoint API coming, so that might do a better job of maintaining breakpoint state.

File: src/mibase.ts [hunk 2]
```diff
@@ -256,13 +258,24 @@ export class MI2DebugSession extends DebugSession {
 	}
 
 	protected threadsRequest(response: DebugProtocol.ThreadsResponse): void {
+		if (!this.miDebugger) {
+			this.sendResponse(response);
```
Comment (WebFreak001) [inline src/mibase.ts]: hm this should also return so the next call doesn't crash

**Pull request merged**